lib/fsm/follow.go

```diff
@@ -22,74 +22,107 @@ import (
 
 	"github.com/gravitational/gravity/lib/storage"
 
-	"github.com/gravitational/trace"
+	"github.com/cenkalti/backoff"
 	"github.com/sirupsen/logrus"
 )
 
 // GetPlanFunc is a function that returns an operation plan.
 type GetPlanFunc func() (*storage.OperationPlan, error)
 
 // PlanEvent represents an operation plan event.
-type PlanEvent interface{}
+type PlanEvent interface {
+	// IsPlanEvent returns true for events representing plan changes.
+	IsPlanEvent() bool
+}
 
-// PlanChanged is sent when plan phases change states.
-type PlanChanged struct {
+// PlanChangeEvent is sent when plan phases change states.
+type PlanChangeEvent struct {
 	// Change is a plan phase change.
 	Change storage.PlanChange
 	// Plan is the resolved operation plan.
 	Plan storage.OperationPlan
```

Comment:
You don't seem to have a use for the plan itself other than the number of phases (aka its `Length`) - do you need to pack the whole plan with every change? Even the length is a static artifact - i.e. it does not change as you follow the plan.